### ISSUE TITLE
Building Error: Wrong checksum for package openjdk-11-jdk-headless

### ISSUE THREAD

**Issue by shibukawa** [GoogleContainerTools/distroless#509]
I am trying the following commands as CONTRIBUTING.md says to create PR for [this issue](https://github.com/GoogleContainerTools/distroless/issues/508). But the following error happens. How can I work around this issue?

```
$ bazel clean --host_force_python=PY2 --curses=no
$ bazel build --host_force_python=PY2 --curses=no //package_manager:dpkg_parser.par
:
INFO: 2 processes: 2 linux-sandbox.
INFO: Build completed successfully, 15 total actions
INFO: Build completed successfully, 15 total actions
$ bazel build --host_force_python=PY2 //experimental/python3:python3_debian10
INFO: Repository package_bundle_debian10 instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule _dpkg_list defined at:
  /home/yoshiki/develop/distroless/package_manager/dpkg.bzl:23:14: in <toplevel>
ERROR: An error occurred during the fetch of repository 'package_bundle_debian10':
   dpkg_parser command failed: Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/yoshiki/develop/distroless/bazel-bin/package_manager/dpkg_parser.par/__main__.py", line 196, in <module>
  File "/home/yoshiki/develop/distroless/bazel-bin/package_manager/dpkg_parser.par/__main__.py", line 80, in main
  File "/home/yoshiki/develop/distroless/bazel-bin/package_manager/dpkg_parser.par/__main__.py", line 118, in download_dpkg
Exception: Wrong checksum for package openjdk-11-jdk-headless.  Expected: a3c04c7f67bd377704e64c7cac9ba14b623e1016489904bd2d64c5e288f2a028, Actual: 864d1df61900a364c8a1cb69c7407ee416e6e40acd3bbb0316a4c414b0475b98
 (/home/yoshiki/develop/distroless/bazel-bin/package_manager/dpkg_parser.par --package-files /home/yoshiki/.cache/bazel/_bazel_yoshiki/341ae8492058161a7bc59a3774bb3b8a/external/debian10_security/file/Packages.json,/home/yoshiki/.cache/bazel/_bazel_yoshiki/341ae8492058161a7bc59a3774bb3b8a/external/debian10/file/Packages.json --packages libc6,base-files,ca-certificates,openssl,libssl1.1,libbz2-1.0,libdb5.3,libffi6,liblzma5,libexpat1,libreadline7,libsqlite3-0,mime-support,netbase,readline-common,tzdata,libgcc1,libgomp1,libstdc++6,zlib1g,libjpeg62-turbo,libpng16-16,liblcms2-2,libfreetype6,fonts-dejavu-core,fontconfig-config,libfontconfig1,libuuid1,openjdk-11-jre-headless,openjdk-11-jdk-headless,dash,libc-bin,libpython2.7-minimal,libpython2.7-stdlib,python2.7-minimal,libmpdec2,libpython3.7-minimal,libpython3.7-stdlib,libtinfo6,libuuid1,libncursesw6,python3-distutils,python3.7-minimal,libcurl4,libgssapi-krb5-2,libicu63,liblttng-ust0,libssl1.1,libunwind8,libuuid1,zlib1g,curl,libcomerr2,libidn2-0,libk5crypto3,libkrb5-3,libldap-2.4-2,libldap-common,libsasl2-2,libnghttp2-14,libpsl5,librtmp1,libssh2-1,libkeyutils1,libkrb5support0,libunistring2,libgnutls30,libgmp10,libhogweed4,libidn11,libnettle6,libp11-kit0,libffi6,libtasn1-6,libsasl2-modules-db,libgcrypt20,libgpg-error0,libacl1,libattr1,libselinux1,libpcre3,libbz2-1.0,liblzma5 --workspace-name package_bundle_debian10)
```

Same error happens even if the command in test.sh:

```
$ bazel build --host_force_python=PY2 --curses=no //...
```

My envirnment is Linux terminal on ChromeOS (Debian 9 on HP Chromebook x360 14 Core i5). Bezel version is the latest 3.1.0.

**Comment by shibukawa** [comment]
WORKSPACES contains the following instruction:

```
dpkg_src(
    name = "debian10_security",
    package_prefix = "https://snapshot.debian.org/archive/debian-security/20200424T130133Z/",
    packages_gz_url = "https://snapshot.debian.org/archive/debian-security/20200424T130133Z//dists/buster/updates/main/binary-amd64/Packages.gz",
    sha256 = "65dab8d4b9c68d6d1726089092940e4c80a7e0d7151dacd9ec6b4cf6bd1f573b",
)
```

This packages_gz_url file contains the following contents:

```
Package: openjdk-11-jdk-headless
Source: openjdk-11
Version: 11.0.7+10-3~deb10u1
Installed-Size: 219724
Maintainer: OpenJDK Team <openjdk@lists.launchpad.net>
Architecture: amd64
Replaces: openjdk-11-jre-headless (<< 11~19-2)
Provides: java-compiler, java-sdk-headless, java10-sdk-headless, java11-sdk-headless, java2-sdk-headless, java5-sdk-headless, java6-sdk-headless, java7-sdk-headless, java8-sdk-headless, java9-sdk-headless
Depends: openjdk-11-jre-headless (= 11.0.7+10-3~deb10u1), libc6 (>= 2.2.5), zlib1g (>= 1:1.1.4)
Suggests: openjdk-11-demo, openjdk-11-source
Conflicts: openjdk-11-jre-headless (<< 11~19-2)
Description: OpenJDK Development Kit (JDK) (headless)
Homepage: http://openjdk.java.net/
Description-md5: 494a52f34d707a832270d6d51a78f511
Multi-Arch: same
Section: java
Priority: optional
Filename: pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb
Size: 215331352
MD5sum: 6171273b954da6da53468fc7176d489f
SHA256: 864d1df61900a364c8a1cb69c7407ee416e6e40acd3bbb0316a4c414b0475b98
```

This SHA256 is in "actual hash" mentioned in the exception. But I couldn't find where is the expected hash defined.

**Comment by floriankoch** [comment]
@shibukawa   the hash is the sha256 in the dpkg_scr rule

**Comment by chanseokoh** [comment]
@floriankoch no, the SHA in the error is for an individual `.deb` package when downloading each package inside `dpkg_parser.py`. `dpkg_parser.py` is comparing the SHA256 of `openjdk11-jdk-headless` declared in `Packages.gz` with the SHA256 computed from the downloaded `.deb`. The error says they don't make. This doesn't make sense. So weird. I wonder if Debian botched their snapshot repo?

BTW, looking into the source of `dpkg_parser.py`, I noticed the words "Expected" and "Actual" are reversed.

`Wrong checksum for package openjdk-11-jdk-headless.  Expected: a3c04c7f67bd377704e64c7cac9ba14b623e1016489904bd2d64c5e288f2a028, Actual: 864d1df61900a364c8a1cb69c7407ee416e6e40acd3bbb0316a4c414b0475b98`

The actual SHA256 locally computed from the downloaded file is `a3c04c7f67bd377704e64c7cac9ba14b623e1016489904bd2d64c5e288f2a028` while the expected SHA256 as declared in `Packages.gz` is `864d1df61900a364c8a1cb69c7407ee416e6e40acd3bbb0316a4c414b0475b98`.

**Comment by Ilhicas** [comment]
Just hitting the same issue here @chanseokoh , I know this internal and all, but its an open repo :D

**Comment by chanseokoh** [comment]
Looks like the Debian server is either very unreliable or might have changed recently to return contents in chunks (as I see it return `206 Partial Content`). I tried manually downloading `openjdk-11-jdk-headless` with `wget`, and I observed the connection was cut off multiple times.

```
$ wget https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.
0.7%2B10-3~deb10u1_amd64.deb
--2020-05-11 11:19:11--  https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7%2B10-3~deb10u1_amd64.deb
Resolving snapshot.debian.org (snapshot.debian.org)... 2001:630:206:4000:1a1a:0:c13e:ca1b, 2001:1af8:4020:b030:deb::185, 193.62.202.27, ...
Connecting to snapshot.debian.org (snapshot.debian.org)|2001:630:206:4000:1a1a:0:c13e:ca1b|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 215331352 (205M)
Saving to: ‘openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb’

openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd6  33%[=================================>                                                                      ]  68.45M  29.6MB/s    in 2.3s    

2020-05-11 11:19:14 (29.6 MB/s) - Connection closed at byte 71770112. Retrying.

--2020-05-11 11:19:15--  (try: 2)  https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7%2B10-3~deb10u1_amd64.deb
Connecting to snapshot.debian.org (snapshot.debian.org)|2001:630:206:4000:1a1a:0:c13e:ca1b|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 215331352 (205M), 143561240 (137M) remaining
Saving to: ‘openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb’

openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd6  53%[++++++++++++++++++++++++++++++++++====================>                                                 ] 109.32M  26.6MB/s    in 1.5s    

2020-05-11 11:19:17 (26.6 MB/s) - Connection closed at byte 114630656. Retrying.

--2020-05-11 11:19:19--  (try: 3)  https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7%2B10-3~deb10u1_amd64.deb
Connecting to snapshot.debian.org (snapshot.debian.org)|2001:630:206:4000:1a1a:0:c13e:ca1b|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 215331352 (205M), 100700696 (96M) remaining
Saving to: ‘openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb’

openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd6  69%[+++++++++++++++++++++++++++++++++++++++++++++++++++++++================>                                ] 142.20M  23.8MB/s    in 1.4s    

2020-05-11 11:19:20 (23.8 MB/s) - Connection closed at byte 149106688. Retrying.

--2020-05-11 11:19:23--  (try: 4)  https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7%2B10-3~deb10u1_amd64.deb
Connecting to snapshot.debian.org (snapshot.debian.org)|2001:630:206:4000:1a1a:0:c13e:ca1b|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 215331352 (205M), 66224664 (63M) remaining
Saving to: ‘openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb’

openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd6 100%[++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++===============================>] 205.36M  29.5MB/s    in 2.1s    

2020-05-11 11:19:26 (29.5 MB/s) - ‘openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb’ saved [215331352/215331352]
```

Fortunately, `wget` automatically handles all of these and retries downloading. When computing the SHA after complete download, the SHA matches the expected.
```
$ sha256sum openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb
864d1df61900a364c8a1cb69c7407ee416e6e40acd3bbb0316a4c414b0475b98  openjdk-11-jdk-headless_11.0.7+10-3~deb10u1_amd64.deb
```

I am not sure if this is a temporary issue on the Debian server or a permanent change in the way they serve contents. If it's a temporary issue, we just need to wait. Otherwise, we might need to fix `dpkg_parser.py`, which will not be trivial to fix.

**Comment by chanseokoh** [comment]
Unlike `wget`, cURL doesn't handle this automatically. But I am more convinced that this is not the fault of the server. cURL clearly reports how many bytes are remaining to read.
```
$ curl -o a.deb https://snapshot.debian.org/archive/debian-security/20200424T130133Z/pool/updates/main/o/openjdk-11/openjdk-11-jdk-headless_11.0.7%2B10-3~deb10u1_amd64.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 37  205M   37 77.8M    0     0  6964k      0  0:00:30  0:00:11  0:00:19 6313k
curl: (18) transfer closed with 133739032 bytes remaining to read
```

**Comment by chanseokoh** [comment]
I've filed a bug report (email sent) to Debian. We'll see.

**Comment by shibukawa** [comment]
https://github.com/shibukawa/distroless/commit/595bfe1918c1ddc855d63c1269c52a9d519aeb71

I have implemented range access if the downloaded content size is shorter than Content-Length header.
I couldn't see the above issue on my branch, but I got 10 minutes timeout and I couldn't finished running successfully. distroless is too big for my computer and/or network.

```
WARNING: Target pattern parsing failed.
ERROR: no such package '@package_bundle//file': dpkg_parser command failed: Timed out (/home/yoshiki/develop/distroless/bazel-bin/package_manager/dpkg_parser.par --package-files /home/yoshiki/.cache/bazel/_bazel_yoshiki/341ae8492058161a7bc59a3774bb3b8a/external/debian_stretch_security/file/Packages.json,/home/yoshiki/.cache/bazel/_bazel_yoshiki/341ae8492058161a7bc59a3774bb3b8a/external/debian_stretch_backports/file/Packages.json,/home/yoshiki/.cache/bazel/_bazel_yoshiki/341ae8492058161a7bc59a3774bb3b8a/external/debian_stretch/file/Packages.json --packages libc6=2.24-11+deb9u4,base-files,ca-certificates,openssl,libssl1.0.2,libssl1.1,libbz2-1.0,libdb5.3,libffi6,libncursesw5,liblzma5,libexpat1,libreadline7,libtinfo5,libsqlite3-0,mime-support,netbase,readline-common,tzdata,libgcc1,libgomp1,libstdc++6,zlib1g,libjpeg62-turbo,libpng16-16,liblcms2-2,libfreetype6,fonts-dejavu-core,fontconfig-config,libfontconfig1,libuuid1,openjdk-8-jre-headless,openjdk-8-jdk-headless,openjdk-11-jre-headless,openjdk-11-jdk-headless,libpython2.7-minimal,python2.7-minimal,libpython2.7-stdlib,dash,libc-bin=2.24-11+deb9u4,libmpdec2,libpython3.5-minimal,libpython3.5-stdlib,python3.5-minimal,libcurl3,libgssapi-krb5-2,libicu57,liblttng-ust0,libssl1.0.2,libunwind8,libuuid1,zlib1g,curl,libcomerr2,libidn2-0,libk5crypto3,libkrb5-3,libldap-2.4-2,libldap-common,libsasl2-2,libnghttp2-14,libpsl5,librtmp1,libssh2-1,libkeyutils1,libkrb5support0,libunistring0,libgnutls30,libgmp10,libhogweed4,libidn11,libnettle6,libp11-kit0,libffi6,libtasn1-6,libsasl2-modules-db,libgcrypt20,libgpg-error0,libacl1,libattr1,libselinux1,libpcre3,libbz2-1.0,liblzma5 --workspace-name package_bundle)
```

**Comment by chanseokoh** [comment]
Oh, wow. Never expected someone will actually implement the Range request fallback. I was thinking Debian should probably fix their server. This is super cool! I'll test your branch on my machine and see if it works.

BTW, do you know the exact location/call that's timing out? Do you who know sets the 10-minute timeout? Is it the 20 retry counts?

**Comment by chanseokoh** [comment]
Maybe Debian fixed their server. Although I don't see any activity on the Debian [bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=960304), connection is not being cut off today.

**Comment by Ilhicas** [comment]
I don't think they did, or I'm getting some old servers to respond to my requests

```sh
Exception: Wrong checksum for package openjdk-11-jdk-headless.  Expected: e906a22dff2533c31ddc1f5e489bdd950458fbd095ce44d6eebf8c366fc71e05, Actual: e154a42aca24eed36f2c874ad3ed517a07187cbefb8b9ad3d036244c2a358ab4
 (/build/bazel-bin/package_manager/dpkg_parser.par --package-files /root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/external/debian_stretch_security/file/Packages.json,/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/external/debian_stretch_backports/file/Packages.json,/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/external/debian_stretch/file/Packages.json --packages libc6=2.24-11+deb9u4,base-files,ca-certificates,openssl,libssl1.0.2,libssl1.1,libbz2-1.0,libdb5.3,libffi6,libncursesw5,liblzma5,libexpat1,libreadline7,libtinfo5,libsqlite3-0,mime-support,netbase,readline-common,tzdata,libgcc1,libgomp1,libstdc++6,bash,zlib1g,libjpeg62-turbo,libpng16-16,liblcms2-2,libfreetype6,fonts-dejavu-core,fontconfig-config,libfontconfig1,openjdk-11-jre-headless,openjdk-11-jdk-headless --workspace-name package_bundle)
```

Haven't tried @shibukawa approach though

**Comment by Ilhicas** [comment]
@shibukawa I confirm your fix is working, thank you for contribution

**Comment by chanseokoh** [comment]
I also tried @shibukawa's fix. However, I think it didn't go through @shibukawa's Range request execution path; on my machine, the Debian server is fully working as before. So I cannot confirm myself if @shibukawa's fix is working.

**Comment by Ilhicas** [comment]
Ive tried the @shibukawa  fix both on local dev machine and on ci system and it works as opposed to calling directly without range request implementation

**Comment by chanseokoh** [comment]
That sounds wonderful. Thank you for confirming. @shibukawa could you put up a PR with the fix?

**Comment by chanseokoh** [comment]
@shibukawa sorry to ping, but could you put up a PR with your fix? We can't take your code ourselves unless you sign and approve contributing the code.

**Comment by reveller** [comment]
@shibukawa I was testing your patch and found a bug at https://github.com/shibukawa/distroless/commit/595bfe1918c1ddc855d63c1269c52a9d519aeb71#diff-4c5ea98df492f916660ff924962cffbfR145-R146 - the req Request defined at line 145 should be used in the call to urlopen in line 146.

Thanks for contributing!

**Comment by shibukawa** [comment]
OK. I will create PR tonight.

**Comment by chanseokoh** [comment]
@Ilhicas the patch had a bug of now making Range requests; I think you were just occasionally lucky and the Debian server did not cut the connection in the middle in those cases.

**Comment by chanseokoh** [comment]
There was [another bug](https://github.com/GoogleContainerTools/distroless/pull/511#issuecomment-630919011). I fixed all the bugs in #511.

**Comment by chanseokoh** [comment]
Fixed by #511.
